File: goa/src/maths/HelpLayer.js
/// <reference path="../cocos2d-typescript-definitions-master/cocos2d/cocos2d-lib.d.ts" />

var xc = xc || {}

xc.HelpLayer = cc.Node.extend({
  _listener: null,
  _graywindow: null,
  _animating: false,
  ctor: function(touchRect, otherRect) {
    this._super()
    if(!otherRect) {
      this._graywindow = this.makePanel(touchRect, cc.rect(0, 0, cc.director.getWinSize().width, cc.director.getWinSize().height))
    } else {
      if((touchRect.x + touchRect.width / 2) < (otherRect.x - otherRect.width)) {
        this._graywindow = this.makePanel(touchRect, cc.rect(0, 0, otherRect.x - otherRect.width / 2, cc.director.getWinSize().height))
        this.makePanel(otherRect, cc.rect(otherRect.x - otherRect.width / 2, 0, cc.director.getWinSize().width - otherRect.x + otherRect.width / 2 , cc.director.getWinSize().height))    
      } else if((otherRect.x + otherRect.width / 2) < (touchRect.x - touchRect.width)) {
        this.makePanel(otherRect, cc.rect(0, 0, touchRect.x - touchRect.width / 2, cc.director.getWinSize().height))
        this._graywindow = this.makePanel(touchRect, cc.rect(touchRect.x - touchRect.width / 2, 0, cc.director.getWinSize().width - touchRect.x + touchRect.width / 2 , cc.director.getWinSize().height))    
      } else if((touchRect.y + touchRect.height / 2) < (otherRect.y - otherRect.height)) {
        this._graywindow = this.makePanel(touchRect, cc.rect(0, 0, cc.director.getWinSize().width, otherRect.y - otherRect.height / 2))
        this.makePanel(otherRect, cc.rect(0, otherRect.y - otherRect.height / 2, cc.director.getWinSize().width, cc.director.getWinSize().height - otherRect.y + otherRect.height / 2 ))    
      } else {
        this.makePanel(otherRect, cc.rect(0, 0, cc.director.getWinSize().width, touchRect.y - touchRect.height / 2))
        this._graywindow = this.makePanel(touchRect, cc.rect(0, touchRect.y - touchRect.height / 2, cc.director.getWinSize().width, cc.director.getWinSize().height - touchRect.y + touchRect.height / 2 ))          
      }          
    }

    this._listener = cc.EventListener.create({
        event: cc.EventListener.TOUCH_ONE_BY_ONE,
        swallowTouches: true,
        onTouchBegan: function (touch, event) {
          var target = event.getCurrentTarget()
          if(target._animating) {
            return true
          }
          if (cc.rectContainsPoint(target._graywindow.getBoundingBox(), touch.getLocation())) {
            return false
          }
          return true
        },
        onTouchMoved: function(touch, event) {
        },
        onTouchEnded: function(touch, event) {
        }
    })
    cc.eventManager.addListener(this._listener, this)
  },
  makePanel: function(windowRect, grayRect) {
    var graywindow = new cc.Sprite(xc.HelpLayer.res.graywindow_png)
    this.addChild(graywindow)
    graywindow.setPosition(windowRect.x, windowRect.y)
    graywindow.setScale(windowRect.width / xc.HelpLayer.GW_WIDTH, windowRect.height / xc.HelpLayer.GW_HEIGHT)

    var bottom = new cc.Sprite(xc.HelpLayer.res.gray_png)
    bottom.setPosition(grayRect.x + grayRect.width / 2, (grayRect.y + (windowRect.y - windowRect.height / 2)) / 2)
    bottom.setScale(grayRect.width / xc.HelpLayer.GRAY_WIDTH, (windowRect.y - windowRect.height / 2 - grayRect.y) / xc.HelpLayer.GRAY_HEIGHT)
    this.addChild(bottom)

    var top = new cc.Sprite(xc.HelpLayer.res.gray_png)
    top.setPosition(grayRect.x + grayRect.width / 2, windowRect.y + windowRect.height / 2 + (grayRect.y + grayRect.height - windowRect.y - windowRect.height / 2) / 2)
    top.setScale(grayRect.width / xc.HelpLayer.GRAY_WIDTH, (grayRect.height + grayRect.y - windowRect.y - windowRect.height / 2) / xc.HelpLayer.GRAY_HEIGHT)
    this.addChild(top)

    var left = new cc.Sprite(xc.HelpLayer.res.gray_png)
    left.setPosition((grayRect.x + windowRect.x - windowRect.width / 2) / 2, windowRect.y)
    left.setScale(((windowRect.x - windowRect.width / 2) - grayRect.x) / xc.HelpLayer.GRAY_WIDTH, windowRect.height / xc.HelpLayer.GRAY_HEIGHT)
    this.addChild(left)

    var right = new cc.Sprite(xc.HelpLayer.res.gray_png)
    right.setPosition(grayRect.x / 2 + windowRect.x + windowRect.width / 2 + (grayRect.width - windowRect.x - windowRect.width / 2) / 2, windowRect.y)
    right.setScale(((grayRect.width + grayRect.x - windowRect.x - windowRect.width / 2)) / xc.HelpLayer.GRAY_WIDTH, windowRect.height / xc.HelpLayer.GRAY_HEIGHT)
    this.addChild(right)    
    return graywindow
  },
  click: function(x, y) {
    this._animating = true
    var touch = new cc.Sprite(xc.HelpLayer.res.touch_png)
    touch.setPosition(x, y)
    touch.setAnchorPoint(0.5, 1.0)
    this.addChild(touch)
    var touchAction = new cc.ScaleBy(0.5, 0.8)
    var callFunc = new cc.CallFunc(function() {
      this.getParent()._animating = false
      this.removeFromParent()
    }, touch)
    var seq = new cc.Sequence(touchAction, cc.delayTime(0.5), touchAction.reverse(), callFunc)
    touch.runAction(seq)
  },
  clickAndDrag: function(startX, startY, endX, endY) {
    this._animating = true
    var touch = new cc.Sprite(xc.HelpLayer.res.touch_png)
    touch.setPosition(startX, startY)
    touch.setAnchorPoint(0.5, 1.0)
    this.addChild(touch)
    var touchAction = new cc.ScaleBy(0.5, 0.8)
    var moveAction = new cc.MoveTo(0.5, endX, endY)
    var callFunc = new cc.CallFunc(function() {
      this.getParent()._animating = false
      this.removeFromParent()
    }, touch)
    var seq = new cc.Sequence(touchAction, cc.delayTime(0.5), moveAction, touchAction.reverse(), callFunc)
    touch.runAction(seq)
    
  },
  clickAndDragCardinalSpline: function(points) {
    this._animating = true
    var touch = new cc.Sprite(xc.HelpLayer.res.touch_png)
    touch.setPosition(points[0])
    touch.setAnchorPoint(0.5, 1.0)
    this.addChild(touch)
    var touchAction = new cc.ScaleBy(0.5, 0.8)
    var pathAction = new cc.CardinalSplineTo(5, points, 0.5)
    var callFunc = new cc.CallFunc(function() {
      this.getParent()._animating = false
      this.removeFromParent()
    }, touch)
    var seq = new cc.Sequence(touchAction, cc.delayTime(0.5), pathAction, touchAction.reverse(), callFunc)
    touch.runAction(seq)
  }
})

xc.HelpLayer.res = {
  graywindow_png: xc.path + "help/graywindow.png", 
  gray_png: xc.path + "help/gray.png",
  touch_png: xc.path + "help/touch.png" 
}

xc.HelpLayer.GW_WIDTH = 800
xc.HelpLayer.GW_HEIGHT = 800
xc.HelpLayer.GRAY_WIDTH = 100
xc.HelpLayer.GRAY_HEIGHT = 100

File: goa/frameworks/runtime-src/Classes/menu/HelpLayer.cpp
//
//  HelpLayer.cpp
//  goa
//
//  Created by Srikanth Talapadi on 18/10/2016.
//
//

#include "HelpLayer.h"

USING_NS_CC;

HelpLayer* HelpLayer::create(Rect touchRect, Rect otherRect) {
    HelpLayer* helpLayer = new (std::nothrow) HelpLayer();
    if(helpLayer && helpLayer->init(touchRect, otherRect)) {
        helpLayer->autorelease();
        return helpLayer;
    }
    CC_SAFE_DELETE(helpLayer);
    return nullptr;
    
}

bool HelpLayer::init(const Rect touchRect, const Rect otherRect) {
    if(!Node::init()) {
        return false;
    }
    if(otherRect.equals(Rect::ZERO)) {
        _grayWindow = makePanel(touchRect, Rect(Director::getInstance()->getVisibleOrigin(), Director::getInstance()->getVisibleSize()));
    } else {
        if((touchRect.origin.x + touchRect.size.width / 2) < (otherRect.origin.x - otherRect.size.width)) {
            _grayWindow = makePanel(touchRect, Rect(0, 0, otherRect.origin.x - otherRect.size.width / 2, Director::getInstance()->getVisibleSize().height));
            makePanel(otherRect, Rect(otherRect.origin.x - otherRect.size.width / 2, 0, Director::getInstance()->getVisibleSize().width - otherRect.origin.x + otherRect.size.width / 2 , Director::getInstance()->getVisibleSize().height));
        } else if((otherRect.origin.x + otherRect.size.width / 2) < (touchRect.origin.x - touchRect.size.width)) {
            makePanel(otherRect, Rect(0, 0, touchRect.origin.x - touchRect.size.width / 2, Director::getInstance()->getVisibleSize().height));
            _grayWindow = makePanel(touchRect, Rect(touchRect.origin.x - touchRect.size.width / 2, 0, Director::getInstance()->getVisibleSize().width - touchRect.origin.x + touchRect.size.width / 2 , Director::getInstance()->getVisibleSize().height));
        } else if((touchRect.origin.y + touchRect.size.height / 2) < (otherRect.origin.y - otherRect.size.height)) {
            _grayWindow = makePanel(touchRect, Rect(0, 0, Director::getInstance()->getVisibleSize().width, otherRect.origin.y - otherRect.size.height / 2));
            makePanel(otherRect, Rect(0, otherRect.origin.y - otherRect.size.height / 2, Director::getInstance()->getVisibleSize().width, Director::getInstance()->getVisibleSize().height - otherRect.origin.y + otherRect.size.height / 2 ));
        } else {
            makePanel(otherRect, Rect(0, 0, Director::getInstance()->getVisibleSize().width, touchRect.origin.y - touchRect.size.height / 2));
            _grayWindow = makePanel(touchRect, Rect(0, touchRect.origin.y - touchRect.size.height / 2, Director::getInstance()->getVisibleSize().width, Director::getInstance()->getVisibleSize().height - touchRect.origin.y + touchRect.size.height / 2 ));
        }
    }
    _listener = EventListenerTouchOneByOne::create();
    _listener->setSwallowTouches(true);
    _listener->onTouchBegan = CC_CALLBACK_2(HelpLayer::onTouchBegan, this);
    _eventDispatcher->addEventListenerWithSceneGraphPriority(_listener, this);
    return true;
}

bool HelpLayer::onTouchBegan(Touch *touch, Event *event) {
    if(_animating) {
        return true;
    }
    if(_grayWindow->getBoundingBox().containsPoint(touch->getLocation())) {
        return false;
    }
    return true;
}

void HelpLayer::click(Vec2 point) {
    _animating = true;
    auto finger = Sprite::create("help/touch.png");
    finger->setPosition(point);
    finger->setAnchorPoint(Vec2::ANCHOR_MIDDLE_TOP);
    addChild(finger);
    auto scaleBy = ScaleBy::create(0.5, 0.8);
    auto hideCallFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::hideFinger, this, finger));
    auto repeatCallFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::click, this, point));
    auto removeCallFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::removeFinger, this, finger));
    finger->runAction(Sequence::create(scaleBy, DelayTime::create(0.5), hideCallFunc, DelayTime::create(4.0), repeatCallFunc, removeCallFunc,  NULL));
}
void HelpLayer::clickTwice(cocos2d::Vec2 point1, cocos2d::Vec2 point2) {
    _animating = true;
    auto finger = Sprite::create("help/touch.png");
    finger->setPosition(point1);
    finger->setAnchorPoint(Vec2::ANCHOR_MIDDLE_TOP);
    addChild(finger);
    auto scaleBy = ScaleBy::create(0.5, 0.8);
    auto moveTo = MoveTo::create(0.5, point2);
    auto callFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::removeFinger, this, finger));
    auto hideCallFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::hideFinger, this, finger));
    auto repeatCallFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::clickTwice, this, point1, point2));
    
    finger->runAction(Sequence::create(scaleBy, DelayTime::create(0.5), scaleBy->reverse(), moveTo, scaleBy, DelayTime::create(0.5), scaleBy->reverse(), hideCallFunc, DelayTime::create(4.0), repeatCallFunc, callFunc, NULL));
}

void HelpLayer::clickAndDrag(Vec2 startPoint, Vec2 endPoint) {
    _animating = true;
    auto finger = Sprite::create("help/touch.png");
    finger->setPosition(startPoint);
    finger->setAnchorPoint(Vec2::ANCHOR_MIDDLE_TOP);
    addChild(finger);
    auto scaleBy = ScaleBy::create(0.5, 0.8);
    auto moveTo = MoveTo::create(0.5, endPoint);
    auto callFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::removeFinger, this, finger));
    auto hideCallFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::hideFinger, this, finger));
    auto repeatCallFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::clickAndDrag, this, startPoint, endPoint));
    
    finger->runAction(Sequence::create(scaleBy, DelayTime::create(0.5), moveTo, scaleBy->reverse(), hideCallFunc, DelayTime::create(4.0), repeatCallFunc, callFunc, NULL));
}

/*
vector contains start point and end point of the each stroke


*/
void HelpLayer::writing(std::vector<cocos2d::Vec2> points)
{
	_animating = true;
	if (points.size() >= 5) {
		auto finger = Sprite::create("help/touch.png");
		finger->setPosition(points.at(0));
		finger->setAnchorPoint(Vec2::ANCHOR_MIDDLE_TOP);
		addChild(finger);
		Vector< FiniteTimeAction * > fta_stroke1;
		
		auto scaleBy = ScaleBy::create(0.5, 0.8);
		fta_stroke1.pushBack(scaleBy);
		fta_stroke1.pushBack(DelayTime::create(0.5));
		auto moveTo = MoveTo::create(0.5, points.at(1));
		fta_stroke1.pushBack(moveTo);
		auto moveTo1 = MoveTo::create(0.5, points.at(2));
		fta_stroke1.pushBack(moveTo1);
		fta_stroke1.pushBack(scaleBy->reverse());
		auto callFunc = CallFunc::create(CC_CALLBACK_0(HelpLayer::removeFinger, this, finger));
		
		auto callFunc1 = CallFunc::create([=]() {
			clickAndDrag(points.at(3), points.at(4));
		});
		fta_stroke1.pushBack(callFunc1);
		fta_stroke1.pushBack(callFunc);
		fta_stroke1.pushBack(DelayTime::create(0.3));
		finger->runAction(Sequence::create(fta_stroke1));
	}
}

void HelpLayer::hideFinger(Node* finger) {
    _animating = false;
    finger->setVisible(false);
}

void HelpLayer::removeFinger(Node* finger) {
    removeChild(finger);
}


static const float GW_WIDTH = 800.0;
static const float GW_HEIGHT = 800.0;
static const float GRAY_WIDTH = 100.0;
static const float GRAY_HEIGHT = 100.0;


Node* HelpLayer::makePanel(Rect windowRect, Rect grayRect) {
    auto grayWindow = Sprite::create("help/graywindow.png");
    addChild(grayWindow);
    grayWindow->setPosition(windowRect.origin);
    grayWindow->setScale(windowRect.size.width / GW_WIDTH, windowRect.size.height / GW_HEIGHT);
    
    auto bottom = Sprite::create("help/gray.png");
    bottom->setPosition(Vec2(grayRect.origin.x + grayRect.size.width / 2, (grayRect.origin.y + (windowRect.origin.y - windowRect.size.height / 2)) / 2));
    bottom->setScale(grayRect.size.width / GRAY_WIDTH, (windowRect.origin.y - windowRect.size.height / 2 - grayRect.origin.y) / GRAY_HEIGHT);
    addChild(bottom);
    
    auto top = Sprite::create("help/gray.png");
    top->setPosition(Vec2(grayRect.origin.x + grayRect.size.width / 2, windowRect.origin.y + windowRect.size.height / 2 + (grayRect.origin.y + grayRect.size.height - windowRect.origin.y - windowRect.size.height / 2) / 2));
    top->setScale(grayRect.size.width / GRAY_WIDTH, (grayRect.size.height + grayRect.origin.y - windowRect.origin.y - windowRect.size.height / 2) / GRAY_HEIGHT);
    addChild(top);

    auto left = Sprite::create("help/gray.png");
    left->setPosition(Vec2((grayRect.origin.x + windowRect.origin.x - windowRect.size.width / 2) / 2, windowRect.origin.y));
    left->setScale(((windowRect.origin.x - windowRect.size.width / 2) - grayRect.origin.x) / GRAY_WIDTH, windowRect.size.height / GRAY_HEIGHT);
    addChild(left);

    auto right = Sprite::create("help/gray.png");
    right->setPosition(Vec2(grayRect.origin.x / 2 + windowRect.origin.x + windowRect.size.width / 2 + (grayRect.size.width - windowRect.origin.x - windowRect.size.width / 2) / 2, windowRect.origin.y));
    right->setScale(((grayRect.size.width + grayRect.origin.x - windowRect.origin.x - windowRect.size.width / 2)) / GRAY_WIDTH, windowRect.size.height / GRAY_HEIGHT);
    addChild(right);
    
    return grayWindow;
    
}

HelpLayer::HelpLayer() :
_animating(false),
_grayWindow(nullptr),
_listener(nullptr)
{
    
}

HelpLayer::~HelpLayer() {
    
}
